File: server/src/main/java/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultMemoryMetrics.java
package com.krillsson.sysapi.core.metrics.defaultimpl;

import com.krillsson.sysapi.core.domain.memory.MemoryLoad;
import com.krillsson.sysapi.core.metrics.MemoryMetrics;
import oshi.hardware.GlobalMemory;
import oshi.hardware.HardwareAbstractionLayer;
import oshi.hardware.VirtualMemory;
import oshi.software.os.OperatingSystem;

public class DefaultMemoryMetrics implements MemoryMetrics {
    private final HardwareAbstractionLayer hal;
    private final OperatingSystem operatingSystem;

    public DefaultMemoryMetrics(HardwareAbstractionLayer hal, OperatingSystem operatingSystem) {
        this.hal = hal;
        this.operatingSystem = operatingSystem;
    }

    @Override
    public MemoryLoad memoryLoad() {
        GlobalMemory memory = hal.getMemory();
        VirtualMemory virtualMemory = memory.getVirtualMemory();
        return new MemoryLoad(operatingSystem.getProcessCount(), virtualMemory.getSwapTotal(), virtualMemory.getSwapUsed(), memory.getTotal(), memory.getAvailable(), usedPercent(memory));
    }

    private int usedPercent(GlobalMemory memory) {
        long free = memory.getAvailable();
        long total = memory.getTotal();
        long used = total - free;
        return (int) (used * 100.0 / (total) + 0.5);
    }

}

File: server/src/main/java/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultMemoryMetrics.kt
package com.krillsson.sysapi.core.metrics.defaultimpl

import com.krillsson.sysapi.core.domain.memory.MemoryLoad
import com.krillsson.sysapi.core.metrics.MemoryMetrics
import oshi.hardware.GlobalMemory
import oshi.hardware.HardwareAbstractionLayer
import oshi.software.os.OperatingSystem

class DefaultMemoryMetrics(
    private val hal: HardwareAbstractionLayer,
    private val operatingSystem: OperatingSystem
) : MemoryMetrics {
    override fun memoryLoad(): MemoryLoad {
        val memory = hal.memory
        val virtualMemory = memory.virtualMemory
        return MemoryLoad(
            operatingSystem.processCount,
            virtualMemory.swapTotal,
            virtualMemory.swapUsed,
            memory.total,
            memory.available,
            usedPercent(memory).toDouble()
        )
    }

    private fun usedPercent(memory: GlobalMemory): Int {
        val free = memory.available
        val total = memory.total
        val used = total - free
        return (used * 100.0 / total + 0.5).toInt()
    }
}